### PR TITLE
Add completion support for bash|zsh|fish

### LINKED ISSUE #42: support fish shell?
Hi, is it possible to also support fish shell?

### LINKED ISSUE #46: "switch completion" command does not work
Hi guys, thank you for this great tool.
Unfortunately there is a bug in the command `shell completion`.

Here what happens:
```bash
$ switch completion
Error: context with name "completion" not found
```

I installed kubeswitch using homebrew, then sourced the switch.sh as described in the docs:
```bash
$ INSTALLATION_PATH=$(brew --prefix switch) && source $INSTALLATION_PATH/switch.sh
```

Here the version:
```bash
$ switch -v
Switch:
		version     : v0.7.1
		build date  : 2022-06-22
		go version  : go1.18.3
		go compiler : gc
		platform    : darwin/amd64
```

Here shell info:
```bash
$ zsh --version
zsh 5.9 (x86_64-apple-darwin21.3.0)
```

It would be great to have this command working, so instead of being forced to manually download an additional file (scripts/_switch.bash) we could source that directly in .zshrc/.bashrc/.bash_profile

### PR DESCRIPTION
**Pull request by aohoy**
Completions were rewritten to make it possible to have completions for all shells.

- `cmd/switcher/switcher.go` was split into subcommands
- new subcommand `switch completion` was introduced in `cmd/switcher/completion.go`
- new subcommands `list-contexts`, `delete-context`, `unset-context` and `current-context` were implemented to make shell script as small as possible
- flags `-u|-d|-c` were moved from shell into binary

Note: fish shell has built-in command `switch` so `kubeswitch` was used for fish 

Fixes #42, fixes #46

### CODE REVIEW

**Comment by danielfoehrKn** [comment]
Thank you!  Will take a look.
Could you also resolve the merge conflicts?

**Comment by avlllo** [comment]
> Could you also resolve the merge conflicts?
Fixed

**Comment by avlllo** [comment]
@danielfoehrKn What do you think about embedding? There will be only one binary which would produce needed shell functions to source them. Final flow be like:
```
source <(switcher init bash —enable-completion)  # here switcher init outputs function switch() from hack/switch/switch.sh with required completions
switch …
```

File: hack/switch/switch.sh [hunk 1]
```diff
@@ -1,591 +1,60 @@
...
-  then
-    $EXECUTABLE_PATH version
-    return
+    EXECUTABLE_PATH="$DEFAULT_EXECUTABLE_PATH"
   fi
 
-  DEBUG_FLAG=''
-  if [ -n "$DEBUG" ]
+  RESPONSE="$($EXECUTABLE_PATH "${opts[@]}")"
+  if [ $? -ne 0 -o -z "$RESPONSE" ]
```
Comment (danielfoehrKn) [inline hack/switch/switch.sh:31]: I would have to check if all commands still work. But if it does it is a really nice refactoring

**Review by danielfoehrKn — COMMENTED** [review]
I am having some questions before I'll try it out. In general, I like the refactoring aspects of it (it is overdue)

File: docs/command_completion.md [hunk 2]
```diff
@@ -3,21 +3,18 @@
...
+### Zsh
+```sh
+echo 'source <(switch completion zsh)' >> ~/.bashrc
 ```
-
-Then source the bash completion script.
\ No newline at end of file
+### Fish
+```sh
+echo 'kubeswitch completion fish | source' >> ~/.config/fish/config.fish
```
Comment (danielfoehrKn) [inline docs/command_completion.md:19]: should be `switch`, right?

**Comment by danielfoehrKn** [comment]
> What do you think about embedding?

I am not quite sure how that is different from your current approach . You could still use the switcher binary to generate the bash|fish|zsh completions and source them. Maybe I am misunderstanding something.

File: docs/command_completion.md (hunk 2 above)
Comment (avlllo) [reply docs/command_completion.md:19]: fish has [built in `switch` command](https://fishshell.com/docs/current/cmds/switch.html) so it is not possible to use `switch` name in fish.

File: pkg/subcommands/set-context/set_context.go [hunk 3]
```diff
@@ -98,8 +98,6 @@ func SetContext(desiredContext string, stores []store.KubeconfigStore, config *t
 				}
 			}
 
-			// print kubeconfig path to std.out -> captured by calling bash script to set KUBECONFIG environment Variable
-			fmt.Print(tempKubeconfigPath)
```
[the remark this answers is not in the document]
Comment (avlllo) [reply pkg/subcommands/set-context/set_context.go:102]: They are actually printed [here](https://github.com/danielfoehrKn/kubeswitch/blob/8e2db42e88b2bfc1d41e838042eb838022f574da/cmd/switcher/context.go#L201). tempKubeconfigPath is returned to caller.

The whole scheme now is following
- switch.sh [calls switcher binary](https://github.com/danielfoehrKn/kubeswitch/blob/8e2db42e88b2bfc1d41e838042eb838022f574da/hack/switch/switch.sh#L30)
- [the binary prints `switched to context “new_context”.`](https://github.com/danielfoehrKn/kubeswitch/blob/8e2db42e88b2bfc1d41e838042eb838022f574da/cmd/switcher/context.go#L205)
- the output is recorded into `RESPONSE` variable
- [if binary’s output matches wildcard `switched to context “*”.`](https://github.com/danielfoehrKn/kubeswitch/blob/8e2db42e88b2bfc1d41e838042eb838022f574da/hack/switch/switch.sh#L39) then everything but what’s in quotes [is stripped from `RESPONSE`](https://github.com/danielfoehrKn/kubeswitch/blob/8e2db42e88b2bfc1d41e838042eb838022f574da/hack/switch/switch.sh#L41) and [used as new `KUBECONFIG`](https://github.com/danielfoehrKn/kubeswitch/blob/8e2db42e88b2bfc1d41e838042eb838022f574da/hack/switch/switch.sh#L57) variable

Corresponding links are embedded above

File: pkg/main.go [hunk 4]
```diff
@@ -163,11 +163,7 @@ func Switcher(stores []store.KubeconfigStore, config *types.Config, stateDir str
 		logger.Warnf("failed to append context to history file: %v", err)
 	}
 
-	// print kubeconfig path to std.out
-	// captured by calling bash script to set KUBECONFIG environment variable
-	fmt.Print(tempKubeconfigPath)
```
[the remark this answers is not in the document]
Comment (avlllo) [reply pkg/main.go:168]: Answered above, they are the same

File: cmd/switcher/namespace.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+		Short:   "Change the current namespace",
+		Long:    `Search namespaces in the current cluster and change to it.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 1 && len(args[0]) > 0 {
+				return ns.SwitchToNamespace(args[0], getKubeconfigPathFromFlag())
+			}
+
+			return ns.SwitchNamespace(getKubeconfigPathFromFlag(), stateDirectory, noIndex)
+		},
+		SilenceErrors: true,
```
Comment (avlllo) [inline cmd/switcher/namespace.go:21]: Need to implement `ValidArgs` function for autocompletion

File: cmd/switcher/context.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+			stores, config, err := initialize()
+			if err != nil {
+				return err
+			}
+
+			kc, err := set_context.SetContext(args[0], stores, config, stateDirectory, noIndex, true)
+			reportNewContext(kc)
+			return err
+		},
+		SilenceUsage: true,
```
Comment (avlllo) [inline cmd/switcher/context.go:87]: Need to implement ValidArgs function for autocompletion

File: cmd/switcher/context.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+		Long:  `Delete context name provided as first argument. KubeContext name has to exist in the current Kubeconfig file.`,
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctxName, err := resolveContextName(args[0])
+			fmt.Println("delete-context", ctxName, args, err)
+			if err != nil {
+				return err
+			}
+			return delete_context.DeleteContext(ctxName)
+		},
```
Comment (avlllo) [inline cmd/switcher/context.go:102]: Need to implement ValidArgs function for autocompletion

File: cmd/switcher/alias.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+		Use:   "rm",
+		Short: "Remove an existing alias",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 || len(args[0]) == 0 {
+				return fmt.Errorf("please provide the alias to remove as the first argument")
+			}
+
+			return alias.RemoveAlias(args[0], stateDirectory)
+		},
+		SilenceErrors: true,
```
Comment (avlllo) [inline cmd/switcher/alias.go:55]: Need to implement ValidArgs function for autocompletion

**Review by avlllo — COMMENTED** [review]
I need to implement several more completion points

**Comment by sjentzsch** [comment]
@avlllo any chance you can continue this great work? :pray:

**Pull request merged**